src/test/memory/test_mmap.rs

```diff
@@ -0,0 +1,349 @@
+    for byte in bb[..page_size()].iter() {
+        assert_eq!(*byte, 0);
+    }
+    validate_shadow_access(&bb[..page_size()])?;
+
+    // Next page should have been overwritten by smolbuf
+    for byte in bb[page_size()..2 * page_size()].iter() {
+        assert_eq!(*byte, 0);
+    }
+    validate_shadow_access(&bb[page_size()..2 * page_size()])?;
+
+    // Last page should be untouched.
+    for byte in bb[2 * page_size()..3 * page_size()].iter() {
+        assert_eq!(*byte, 2);
+    }
+    validate_shadow_access(&bb[2 * page_size()..])?;
+
+    // Validate Shadow access of the whole buffer (which crosses mmap'd regions) at once.
+    validate_shadow_access(&bb[..])?;
+
+    // Unmap allocated memory
+    let rv = unsafe { libc::munmap(bigbuf, 3 * page_size()) };
+    nix::errno::Errno::result(rv)?;
+
+    Ok(())
+}
+
+// Exercises features used by libpthread when allocating a stack.
+// This includes:
+//   * using PROT_NONE (and then following up with an mprotect to make it accessible).
+//   * using MAP_STACK.
+fn test_mmap_prot_none_mprotect() -> Result<(), Box<dyn Error>> {
+    let size = 8 * 1 << 20;
+
+    // Initially mapped with PROT_NONE, making it inaccessible.
+    let buf_ptr = unsafe {
+        libc::mmap(
+            0 as *mut libc::c_void,
+            size,
+            libc::PROT_NONE,
+            libc::MAP_PRIVATE | libc::MAP_ANONYMOUS | libc::MAP_STACK,
+            -1,
+            0,
+        )
+    };
+    test_utils::assert_true_and_errno(buf_ptr != libc::MAP_FAILED);
+
+    // Update protections to make it accessible.
+    let rv = unsafe { libc::mprotect(buf_ptr, size, libc::PROT_READ | libc::PROT_WRITE) };
+    test_utils::assert_true_and_errno(rv == 0);
+
+    let mut buf = unsafe { std::slice::from_raw_parts_mut::<u8>(buf_ptr as *mut u8, size) };
+    // Validate that it's accessible both to the plugin and to Shadow.
+    init_buf(&mut buf);
+    validate_shadow_access(&buf)?;
+
+    // Unmap allocated memory
+    let rv = unsafe { libc::munmap(buf_ptr, size) };
+
+    nix::errno::Errno::result(rv)?;
+
+    Ok(())
+}
+
+fn test_mmap_file_low() -> Result<(), Box<dyn Error>> {
+    test_mmap_file(0)
+}
+
+fn test_mmap_file_high32() -> Result<(), Box<dyn Error>> {
+    test_mmap_file(1 << 20)
+}
+
+// mmap2(2) says highest supported file size is 2**44. Use an offset a bit smaller than
+// that. On a 64-bit system, presumably mmap can handle higher than that, but it's
+// unclear what the limit is. Assume it can handle at least as much as mmap2.
+fn test_mmap_file_high64() -> Result<(), Box<dyn Error>> {
+    test_mmap_file(1 << 43)
+}
+
+// mmap2(2) says highest supported file size is 2**44. Use an offset a bit smaller than
+// that. On a 64-bit system, presumably mmap can handle higher than that, but it's
+// unclear what the limit is. Assume it can handle at least as much as mmap2.
+fn test_mmap2_file_high64() -> Result<(), Box<dyn Error>> {
+    test_mmap_file(1 << 43 / 4096)
+}
```

Comment:
I don't think this test makes sense if not using `mmap2`, since the `4096` is only relevant to `mmap2` where the offset is in 4096-byte chunks. So this test can be removed.